homeassistant/components/vesync/switch.py

```diff
@@ -1,50 +1,44 @@
 """Support for Etekcity VeSync switches."""
 import logging
-import voluptuous as vol
-from homeassistant.components.switch import (SwitchDevice, PLATFORM_SCHEMA)
-from homeassistant.const import (CONF_USERNAME, CONF_PASSWORD)
-import homeassistant.helpers.config_validation as cv
-
+from homeassistant.components.switch import (SwitchDevice)
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from .common import CONF_SWITCHES, async_add_devices
+from .const import VS_DISCOVERY, VS_DISPATCHERS
 
 _LOGGER = logging.getLogger(__name__)
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
-    vol.Required(CONF_USERNAME): cv.string,
-    vol.Required(CONF_PASSWORD): cv.string,
-})
-
-
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the VeSync switch platform."""
-    from pyvesync_v2.vesync import VeSync
+ENERGY_UPDATE_INT = 21600
 
-    switches = []
+DOMAIN = 'vesync'
 
-    manager = VeSync(config.get(CONF_USERNAME), config.get(CONF_PASSWORD))
 
-    if not manager.login():
-        _LOGGER.error("Unable to login to VeSync")
-        return
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up switches."""
+    async def async_discover(dev_len):
+        if dev_len > 0:
+            dev_slice = hass.data[DOMAIN][CONF_SWITCHES][-dev_len:]
+            for dev in dev_slice:
+                await _async_setup_entity(dev, async_add_entities)
 
-    manager.update()
+    disp = async_dispatcher_connect(
+        hass, VS_DISCOVERY.format(CONF_SWITCHES), async_discover)
 
-    if manager.devices is not None and manager.devices:
-        if len(manager.devices) == 1:
-            count_string = 'switch'
-        else:
-            count_string = 'switches'
+    hass.data[DOMAIN][VS_DISPATCHERS].append(disp)
+    await async_add_devices(
+        hass,
+        async_add_entities,
+        hass.data[DOMAIN][CONF_SWITCHES],
+        _async_setup_entity
+    )
+    return True
 
-        _LOGGER.info("Discovered %d VeSync %s",
-                     len(manager.devices), count_string)
 
-        for switch in manager.devices:
-            switches.append(VeSyncSwitchHA(switch))
-            _LOGGER.info("Added a VeSync switch named '%s'",
-                         switch.device_name)
-    else:
-        _LOGGER.info("No VeSync devices found")
-
-    add_entities(switches)
+async def _async_setup_entity(device, async_add_entities):
+    """Check if device is online and add entity."""
+    async_add_entities(
+        [VeSyncSwitchHA(device)],
```

Comment:
Please add all known entitites in a single call to `async_add_entities`. Don't call `async_add_entities` once per entity.